adb/adb_commands.py

```diff
@@ -171,19 +172,39 @@ def Pull(self, device_filename, dest_file=None, timeout_ms=None):
     Returns:
       The file data if dest_file is not set.
     """
-    if isinstance(dest_file, basestring):
-      dest_file = open(dest_file, 'w')
-    elif not dest_file:
-      dest_file = cStringIO.StringIO()
-    connection = self.protocol_handler.Open(
-        self.handle, destination='sync:',
-        timeout_ms=timeout_ms)
-    self.filesync_handler.Pull(connection, device_filename, dest_file)
-    connection.Close()
-    # An empty call to cStringIO.StringIO returns an instance of
-    # cStringIO.OutputType.
-    if isinstance(dest_file, cStringIO.OutputType):
-      return dest_file.getvalue()
+
+    filemode, _, _= self.Stat(device_filename)
+
+    if stat.S_ISDIR(filemode):
+      if dest_file is None:
+        raise ValueError("Must specify dest_file when pulling a directory")
```

Comment:
Alternatively, if they call without dest_file, can we return a map of StringIO objects?

dest_file = collections.defaultdict(cStringIO.StringIO)
...
self.Pull(..., os.path.join(dest_file, device_file.filename) if isinstance(dest_file, basestring) else dest_file[device_file.filename])